Support Flatpak preinstallation as part of a DNF install

owtaylor opened the pull request:
Add new functionality for running Flatpak preinstallation after installing the base system. It's initially implemented only for the DNF payload.
    
Flatpak installation is done as an extra "side" payload that the payloads service calls after the base payload to install additional content. The base payload provides the list of Flatpak refs that should be installed from the side payload and the Flatpaks are installed from a location that is determined from the base payload's primary source - if the source is a local or remote install tree we look for a OCI image layout there, if the base payload is the network (CDN, closest mirror), we install directly from the configured Flatpak remote registry.

 * Preinstallation is done using new Flatpak features that we expect to have merged upstream soon
    - [/etc/flatpak/preinstall.d](https://github.com/flatpak/flatpak/issues/5579)
    - [OCI sideload repositories](https://github.com/flatpak/flatpak/issues/5983)
 * For DNF payloads, we determine what Flatpaks to preinstall by looking for `flatpak-preinstall(REF)` RPM provides rather than directly examining the /etc/flatpak/preinstall.d of the installed system - this allow us to do accurate pre-installation computation of the required install size.
 * For http/ftp payloads we mirror the OCI image layout locally before running the installation.

I have not worked on tests yet, to wait on final agreement on the structure of the code.

## TODO

* [ ] **Tests** pass and cover your changes. You can [run tests locally manually](https://anaconda-installer.readthedocs.io/en/latest/testing.html), or have them run as part of the PR. A team member will have to enable tests manually after inspecting the PR.

* [ ] [**Release notes**](https://anaconda-installer.readthedocs.io/en/latest/contributing.html#release-notes) and **docs** if the PR adds something major or changes existing behavior.

pep8speaks commented:
Hello @owtaylor! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

























* In the file [`pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py`](https://github.com/rhinstaller/anaconda/blob/f2f1cb26b1279b891c9f04010d3f7a9fb76b2f2a/pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py):

> [Line 42:1](https://github.com/rhinstaller/anaconda/blob/f2f1cb26b1279b891c9f04010d3f7a9fb76b2f2a/pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py#L42): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 43:1](https://github.com/rhinstaller/anaconda/blob/f2f1cb26b1279b891c9f04010d3f7a9fb76b2f2a/pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py#L43): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 52:100](https://github.com/rhinstaller/anaconda/blob/f2f1cb26b1279b891c9f04010d3f7a9fb76b2f2a/pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py#L52): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (126 > 99 characters)
> [Line 53:100](https://github.com/rhinstaller/anaconda/blob/f2f1cb26b1279b891c9f04010d3f7a9fb76b2f2a/pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py#L53): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (125 > 99 characters)
> [Line 54:100](https://github.com/rhinstaller/anaconda/blob/f2f1cb26b1279b891c9f04010d3f7a9fb76b2f2a/pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py#L54): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (134 > 99 characters)
> [Line 55:100](https://github.com/rhinstaller/anaconda/blob/f2f1cb26b1279b891c9f04010d3f7a9fb76b2f2a/pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py#L55): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (129 > 99 characters)
> [Line 130:100](https://github.com/rhinstaller/anaconda/blob/f2f1cb26b1279b891c9f04010d3f7a9fb76b2f2a/pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py#L130): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 99 characters)
> [Line 131:100](https://github.com/rhinstaller/anaconda/blob/f2f1cb26b1279b891c9f04010d3f7a9fb76b2f2a/pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py#L131): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 99 characters)



* In the file [`pyanaconda/modules/payloads/payload/flatpak/source.py`](https://github.com/rhinstaller/anaconda/blob/f2f1cb26b1279b891c9f04010d3f7a9fb76b2f2a/pyanaconda/modules/payloads/payload/flatpak/source.py):

> [Line 50:1](https://github.com/rhinstaller/anaconda/blob/f2f1cb26b1279b891c9f04010d3f7a9fb76b2f2a/pyanaconda/modules/payloads/payload/flatpak/source.py#L50): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 165:24](https://github.com/rhinstaller/anaconda/blob/f2f1cb26b1279b891c9f04010d3f7a9fb76b2f2a/pyanaconda/modules/payloads/payload/flatpak/source.py#L165): [E713](https://duckduckgo.com/?q=pep8%20E713) test for membership should be 'not in'
> [Line 393:32](https://github.com/rhinstaller/anaconda/blob/f2f1cb26b1279b891c9f04010d3f7a9fb76b2f2a/pyanaconda/modules/payloads/payload/flatpak/source.py#L393): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

cgwalters commented:
> For DNF payloads, we determine what Flatpaks to preinstall by looking for flatpak-preinstall(REF) RPM provides

Funky. This seems like the biggest thing that would mean this wouldn't work for bootc installs as is.

> this allow us to do accurate pre-installation computation of the required install size.

This also relates to https://github.com/containers/bootc/issues/833 (part of bootc's LBIs).

And this overall effort also kind of relates to https://github.com/rhinstaller/anaconda/discussions/5197 - specifically where I was thinking that it makes sense to restructure things in the general case so that we do an install to a disk, and then we parse the target root and make further decisions based on that for the bootc flow at least.

When we do that, then the need for a lot of magic goes away and we can just parse `/etc/flatpak` - or even better, *run code* from the target root (i.e. the flatpak-being-installed as opposed to flatpak-in-anaconda) - same concerns as bootc has.

The thing I would note here is yes (short of the "artifact pattern" I brought up for bootc) we do lose the space calculation but...what I could imagine doing in the general case is having a "minimum size" as an annotation on the container image at least for bootc.

Yes it's obviously nice to have precise byte-level info but eh...as long as it's *roughly* correct that seems sufficient in the general case.

owtaylor commented:
> > For DNF payloads, we determine what Flatpaks to preinstall by looking for flatpak-preinstall(REF) RPM provides
> 
> Funky. This seems like the biggest thing that would mean this wouldn't work for bootc installs as is

The list of refs to install comes from a virtual function in the payload, so for rpm-ostree/bootc, you do one of two things:

 A) Download the contents of /etc/flatpak/preinstall.d to find the same information (this would be pretty easy with an ostree repository, but not so much with a bootc image as OCI or containers/storage). The virtual provides are just a duplicate of this information for the installer's convenience.
 B) Just assume that every Flatpak on the install media will be installed - presumably the set of Flatpaks that was bundled onto the install media is the set that the image preinstalls. [this requires a minor change to the current code so `None` represents "install all Flatpaks from the media" rather than being the same as `[]`]

What is hard to support with the current framework the case where we are installing directly from registry.redhat.io and then that image pulls in some set of Flatpaks for preinstallation also from registry.redhat.io.

> The thing I would note here is yes (short of the "artifact pattern" I brought up for bootc) we do lose the space calculation but...what I could imagine doing in the general case is having a "minimum size" as an annotation on the container image at least for bootc.
> 
> Yes it's obviously nice to have precise byte-level info but eh...as long as it's roughly correct that seems sufficient in the general case.

Yeah, in the bootc case where the image is bound to set of installed containers, that could be done

For DNF case is that until software selection is done, we have no idea whether this is a minimal server install, or a desktop install with runtime+firefox+thunderbird, and reserving several gigabytes extra of space for every RHEL install just in case seemed really ugly.

cgwalters commented:
> Download the contents of /etc/flatpak/preinstall.d to find the same information (this would be pretty easy with an ostree repository, but not so much with a bootc image as OCI or containers/storage

Side note: [zstd:chunked](https://github.com/containers/storage/blob/main/docs/containers-storage-zstd-chunked.md) makes it possible to efficiently access a subset of the files from a remote container image. Not saying we should rely on it or that it's even a good idea, just noting it.

> Just assume that every Flatpak on the install media will be installed 

I would implement this as "installer generation copies /etc/flatpak/preinstall.d from the target to the installer ISO". I can imagine use cases where the installer environment has a flatpak that we don't want to be present on the target.

> For DNF case is that until software selection is done, we have no idea whether this is a minimal server install, or a desktop install with runtime+firefox+thunderbird, and reserving several gigabytes extra of space for every RHEL install just in case seemed really ugly.

I was thinking of the "minimum space label" in the bootc context, not the RPM context. But I could totally imagine a `Provides: minimum-root-size(10G)` or something that would be attached to something like `fedora-release-workstation` or so.

owtaylor commented:
> > Download the contents of /etc/flatpak/preinstall.d to find the same information (this would be pretty easy with an ostree repository, but not so much with a bootc image as OCI or containers/storage
> 
> Side note: [zstd:chunked](https://github.com/containers/storage/blob/main/docs/containers-storage-zstd-chunked.md) makes it possible to efficiently access a subset of the files from a remote container image. Not saying we should rely on it or that it's even a good idea, just noting it.

Even so, you'd need to handle layering, etc. Would be messy. Maybe if it's implemented it inside 

> > Just assume that every Flatpak on the install media will be installed
> 
> I would implement this as "installer generation copies /etc/flatpak/preinstall.d from the target to the installer ISO". I can imagine use cases where the installer environment has a flatpak that we don't want to be present on the target.

The case where the installer is ostree based and needs a Flatpak (Firefox say) at install time and we want to use the image we're shipping for preinstallation rather than baking the Flatpak into the installer's ostree is tricky - currently installing an image as a Flatpak involves importing it from scratch into the target's /var/lib/flatpak/repo - while doing that at installer boot time might work it's definitely ugly.

I can wave my hands here and say something about "composefs support for Flatpak" and "common object store" - but quite a bit of non-trivial work would be involved :-)

cgwalters commented:
Are you trying to handle a case where there's a single ISO that can support server or desktop install (like the current RHEL 9 ISO) but where flatpaks can exist on the installer ISO but whether or not to install those is a dynamic determination?

Rereading, it seems you must be, which I can understand. But it sure does ratchet the complexity up versus shipping a separate opinionated desktop ISO.

I know this may seem somewhat radical but did you consider at all making the payload metadata here run via a OCI and not by RPM? And by this I mean something like shipping a container image like `registry.redhat.io/rhel10/rhel-workstation-reference` that is just a reference holder which points to say a comps group and a set of flatpaks. This would drop the magical `Provides` that link RPM :arrow_right: flatpak. And more generally, make RPM subordinate to containers, which is how I think things should be going.

owtaylor commented:
> Are you trying to handle a case where there's a single ISO that can support server or desktop install (like the current RHEL 9 ISO) but where flatpaks can exist on the installer ISO but whether or not to install those is a dynamic determination?

Yes - in RHEL 10, Firefox is provided *only* by Flatpak, so the installer ISO has to install the Flatpak iff. the user has selected the appropriate software. (Also Thunderbird, but less important.)

> Rereading, it seems you must be, which I can understand. But it sure does ratchet the complexity up versus shipping a separate opinionated desktop ISO.

Unfortunately, "server with GUI" is still a thing - we need to treat adding a desktop interface as something that you can optionally add onto a server install.

> I know this may seem somewhat radical but did you consider at all making the payload metadata here run via a OCI and not by RPM? And by this I mean something like shipping a container image like `registry.redhat.io/rhel10/rhel-workstation-reference` that is just a reference holder which points to say a comps group and a set of flatpaks. This would drop the magical `Provides` that link RPM ➡️ flatpak. And more generally, make RPM subordinate to containers, which is how I think things should be going.

Don't quite follow how that would work in detail, but generally trying to keep this a fairly minimal change to how we install RHEL. A) system image defines what Flatpaks should be installed B) Anaconda invokes Flatpak to make that happen before first boot. The "magical" Provides links just allow us to fast-forward and figure out what Fltapaks should be installed as soon as the DNF transaction is resolved.

cgwalters commented:
> Don't quite follow how that would work in detail, but generally trying to keep this a fairly minimal change to how we install RHEL. 

I don't think in practice there's a way to avoid that. Moving some apps into flatpaks has a large impact on UX both at install time and postinstall. In particular one thing I have learned the hard way is a lot of important customers do disconnected installs - we have to support that super well. And they do it for good reason (being directly connected to the internet brings large risks). Moving things into flatpaks from RPMs means that these people *will* have to figure out how to mirror the containers in addition to RPMs.

This problem was so painful for OpenShift that we moved everything (except the base RHCOS disk images) into a [release image](https://amd64.ocp.releases.ci.openshift.org/) which is just a reference to 100 other containers, including the operating system content.

So people who want to disconnected installs just need to know how to mirror containers (not RPMs!).

Yes, my suggestion of having the metadata for what to install be referenced via a container image (or perhaps better for this use case an OCI artifact) may seem radical but since containers are already involved (via flatpak) I don't think it actually is too radical. Is it hackier to reference containers from an RPM or RPMS + containers from another container (or OCI artifact)? I think the latter is less hacky.

It's of course worth noting that I have a pretty strong interest in my own giant change for how RHEL is installed via bootc, where RPM is not the technical root, but I do want flatpak preinstalls to work for that too. If we structured the input to both as a container image (or artifact) that would work more smoothly I think.

cgwalters commented:
BTW I should clarify that I am not an Anaconda developer so my comments here should not be construed as blocking, merely a discussion.

bcl commented on pyanaconda/modules/payloads/payload/flatpak/flatpak.py:
```diff
@@ -0,0 +1,112 @@
+#
+# Kickstart module for Live OS payload.
```
Copy paste typo :)

jkonecny12 replied:
Fixed already.

jkonecny12 commented:
Hi @owtaylor, we did some changes in the code to enable pylint changes and we also changed the `master` -> `main` default branch (that should work out of the box hopefully). I mean, you need to rebase to resolve these conflicts but hopefully these shouldn't be painful to resolve.

jkonecny12 commented:
@cgwalters, I see your points and honestly I'm thinking how to involve or not involve DNF into the loop. I know the current implementation is attached to DNF payload which works for standard installations but I'm not sure how it will be handled by bootc so we need to find this out.

After we have this initial version set, I'm planning to schedule a meeting between you, @owtaylor, us and DNF team to figure the ideal solution for RHEL-11. We should think it through to find out what is the best solution (and I'm not saying that this is not the one). In general it would be great to find solution which would be integrating all the pieces together.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/flatpak.py:
```diff
@@ -0,0 +1,112 @@
+#
+# Kickstart module for Live OS payload.
+#
+# Copyright (C) 2019 Red Hat, Inc.
```
Also the year should be fixed.

owtaylor replied:
Fixed.

jkonecny12 reviewed (commented):
Just a partial review right now. So far I wasn't able to find any big issue but the whole concept needs a bit of discussion in the team. I mean the whole solution of attaching this to DNF will not work for other types of payload.

I'll finish this review on January. Thanks for the PR!

owtaylor commented:
> > Don't quite follow how that would work in detail, but generally trying to keep this a fairly minimal change to how we install RHEL.
> 
> I don't think in practice there's a way to avoid that. Moving some apps into flatpaks has a large impact on UX both at install time and postinstall. In particular one thing I have learned the hard way is a lot of important customers do disconnected installs - we have to support that super well. And they do it for good reason (being directly connected to the internet brings large risks). Moving things into flatpaks from RPMs means that these people _will_ have to figure out how to mirror the containers in addition to RPMs.
>
> This problem was so painful for OpenShift that we moved everything (except the base RHCOS disk images) into a [release image](https://amd64.ocp.releases.ci.openshift.org/) which is just a reference to 100 other containers, including the operating system content.

This PR handles disconnected installation whether from a DVD or hard drive install, or from an airgapped ftp/http server - the install tree contains an OCI image layout with all the Flatpaks.

It does *not*, of course, handle disconnected updates - that's outside the scope of Anaconda. It's possible to handle in the same way as this PR works, create an OCI image layout with the latest versions of the Flatpaks (using `skopeo copy`) and then `flatpak update --sideload-repo=<path to  repo>`. But yes, that requires customer adaptation. (We actually *are* going to offer a Firefox RPM for RHEL 10, not installed by default, and marked as deprecated to smooth out the transition for customers.)
 
> So people who want to disconnected installs just need to know how to mirror containers (not RPMs!).
> 
> Yes, my suggestion of having the metadata for what to install be referenced via a container image (or perhaps better for this use case an OCI artifact) may seem radical but since containers are already involved (via flatpak) I don't think it actually is too radical. Is it hackier to reference containers from an RPM or RPMS + containers from another container (or OCI artifact)? I think the latter is less hacky.
> 
> It's of course worth noting that I have a pretty strong interest in my own giant change for how RHEL is installed via bootc, where RPM is not the technical root, but I do want flatpak preinstalls to work for that too. If we structured the input to both as a container image (or artifact) that would work more smoothly I think.

I think this is very compatible with bootc - the current approach is intended to be quite aligned with the way you are handling LBI's. If we wanted to pull the set of Flatpaks to preinstall out from the filesystem tree and expose it in container metadata when creating the bootc image, we could definitely do that - that would give the same ability we have for DNF to preflight space calculations.

But I don't see a need to break the integration with DNF software selection that we have in this patch - different install method, different integration with the base payload.

owtaylor commented:
> Just a partial review right now. So far I wasn't able to find any big issue but the whole concept needs a bit of discussion in the team. I mean the whole solution of attaching this to DNF will not work for other types of payload.
> 
> I'll finish this review on January. Thanks for the PR!

I've rebased it to origin/main, dealing with the import reordering, and checked that the result passes ruff. I have *not* actually tried an install with the rebase (need to pack for the holidays), but hopefully the imports are at least very close to correct. :smiley:

owtaylor commented:
> @cgwalters, I see your points and honestly I'm thinking how to involve or not involve DNF into the loop. I know the current implementation is attached to DNF payload which works for standard installations but I'm not sure how it will be handled by bootc so we need to find this out.

The integration with DNF here is pluggable, and adapting to bootc should not be hard. I was hoping to actually just go and implement that to clarify how it would work, but figuring out how to create a modified anaconda+bootc install image is more than I have cycles for in the short term. 

> After we have this initial version set, I'm planning to schedule a meeting between you, @owtaylor, us and DNF team to figure the ideal solution for RHEL-11. We should think it through to find out what is the best solution (and I'm not saying that this is not the one). In general it would bhave a chance to work one great to find solution which would be integrating all the pieces together.

That sounds good. Basic bootc support is really more something we'll need for 10.x but should be < 100 lines of code on top of this.

jkonecny12 commented:
/kickstart-test --testtype smoke

hanthor commented:
Is anaconda-webui on the roadmap for 10.x?

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,433 @@
+    return collection, "/".join(parts)
+
+
+class NoSourceError(Exception):
```
If this exception could reach UI (other side of DBus) it should be placed in `modules/common/errors/payload.py` and it needs to have `@dbus_error` decorator. The decorator will change the exception to the DBus error and you can create the exception on the UI side.

Also there is already `SourceError` exception which I think could be used here with a details in the message.

If it is used only internally then I would recommend it to move to `payload/flatpak/errors.py` anyway.

owtaylor replied:
It can't reach the UI with the current code structure - both places where it could be thrown are caught in FlatpakManager. Currently we log and ignore to try and make things somewhat robust if a user hasn't set things up properly for a DVD/FTP/HTTP install - you at least get everything else installed. But yes, there is some argument for erroring out as well, which could encourage this to be D-Bus error that just throws through. Do you have an opinion on log+ignore or error out?

jkonecny12 replied:
If it is just about printing error than it should be fine to just use `log.error` without breaking the installation completely.

jkonecny12 replied:
I'm changing the `NoSourceError` to existing `SourceSetupError`.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,433 @@
+
+    @property
+    def labels(self):
+        return self.config_json["config"]["Labels"]
```
It would be great if we can just use a tool for this data grabbing. The tool could be part of the flatpak package and be inline when the metadata will change. In general, if possible we are trying to avoid having too much dependency on someones metadata in Anaconda because these could change in time and most of the time the authors don't know we are using it.
From a quick look it seems that half of the `source.py` code is about metadata parsing and reading.

This is probably for future improvement but I would like to know what you are thinking about it.

jkonecny12 replied:
If it is harder to decouple I guess it is fine then.

jkonecny12 replied:
This is only a json reader, we are not really doing magic here, just taking a json and reading values from it. Also if I understand it correctly these are standardized OCI labels which are used for OCI containers everywhere. I honestly think that we will start doing something similar to Image Mode too. 

It would be nice to have a tooling for that but I don't think it is requirement right now.

jkonecny12 replied:
Also given the fact that the labels are OCI, if there will be backward compatibility change (in how these values are read) Anaconda is not the biggest problem the OCI containers will be facing. The specific values which are used here and by Flatpak may change (that would also break stuff) but that is not hard to fix and tool for us would be in the same situation as our code.

jkonecny12 commented on pyanaconda/timezone.py:
```diff
@@ -24,11 +24,11 @@
 import datetime
 import time
-import zoneinfo
 from collections import OrderedDict
 from functools import cache
 
 import langtable
+import zoneinfo
```
Is this expected change in this commit? Looks to me like a bug from the rebase.

owtaylor replied:
Yeah that snuck in accidentally - I ran `ruff --fix` at some point, though I don't know why it would do that move. Will fix that in the next iteration.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py:
```diff
@@ -0,0 +1,280 @@
+# we start any threads. Setting these variables will be harmless if
+# we aren't actually using Flatpak.
+
+# pylint: disable=environment-modify
+os.environ["FLATPAK_DOWNLOAD_TMPDIR"] = os.path.join(conf.target.system_root, "var/tmp")
+# pylint: disable=environment-modify
+os.environ["FLATPAK_CONFIG_DIR"] = os.path.join(conf.target.system_root, "etc/flatpak")
+# pylint: disable=environment-modify
+os.environ["FLATPAK_OS_CONFIG_DIR"] = os.path.join(conf.target.system_root, "usr/share/flatpak")
+# pylint: disable=environment-modify
+os.environ["FLATPAK_SYSTEM_DIR"] = os.path.join(conf.target.system_root, "var/lib/flatpak")
```
Are these safe to be used on a running system?

We support a lot of variety environments and one of it is Live or image installation. It would be bad if we break the user system after installation because we have changed the environment variables of the system.

owtaylor replied:
These environment variables changes should only affect anaconda and it's child processes, right?

owtaylor replied:
> Last I checked setting `os.environ` in a multithreaded process was still generally unsafe, and Anaconda is definitely one. xref https://sourceware.org/bugzilla/show_bug.cgi?id=15607

The import happens before threads are started. I originally had it very explicitly happening at process initialization: https://github.com/owtaylor/anaconda/blob/flatpak-preinstallation-try-1/pyanaconda/modules/payloads/__main__.py but @jkonecny12 didn't like Flatpak code leaking into a generic place.

jkonecny12 replied:
I see your point. In that case you can move it to `payloads/flatpak/flatpak.py`. My point was mainly to move it out of the shared payloads code which is imported to every payload.

jkonecny12 reviewed (commented):
First round of the review from me. Some notes below.

owtaylor commented on pyanaconda/modules/payloads/payloads.py:
```diff
@@ -89,6 +93,14 @@ def active_payload(self):
+
+        if self._active_payload.needs_flatpak_side_payload():
+            payload = self.create_payload(PayloadType.FLATPAK)
+            assert isinstance(payload, FlatpakModule)
```
This is for the purposes of type checking (I'm using vscode with pylance) - structurally we *know* the result is a FlatpakModule but tooling doesn't. I would say this is the idiomatic way to do it, but it could be also written:

```
payload = typing.cast(self.create_payload(PayloadType.FLATPAK), FlatpakModule)
```

Would that be better?

jkonecny12 replied:
I don't think we want to change code because of tooling. I'm not strictly against the second solution but it would have to be with comment explaining why this is needed.

owtaylor replied:
Something like:

``` python
# create_payload() creates a different subclass of PayloadBase depending on the argument;
# typing.cast() the result, since we are going to call FlatpakModule methods
payload = typing.cast(self.create_payload(PayloadType.FLATPAK), FlatpakModule)
```

I can also just drop it... it's not the only place in Anaconda that shows type warnings in vscode, obviously.

[ having typechecking in your editor really makes development in Python much more productive, highly recommend ]

owtaylor commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,433 @@
+                    if runtime_ref not in result:
+                        result.append(runtime_ref)
+            except (NoSectionError, KeyError):
+                pass
```
I don't think so - it's not an error condition, it's just an Runtime (common), or an application that doesn't use a runtime (logically possible, IIRC, but exceedingly rare.)

owtaylor commented:
I pushed a new version that hopefully addresses all your comments. I left the fixes in a fixup patch for ease of re-review - I can squash and repush if they look OK.

jkonecny12 commented:
The changeset looks good to me. Feel free to squash this. And thanks for this, it really simplifies review ;)

jkonecny12 commented on pyanaconda/modules/payloads/payloads.py:
```diff
@@ -142,6 +153,10 @@ def calculate_required_space(self):
         if self.active_payload:
             total += self.active_payload.calculate_required_space()
+            if self._flatpak_side_payload:
+                self._flatpak_side_payload.set_sources(self.active_payload.get_sources())
```
I think this should be `self.active_payload.sources`.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py:
```diff
@@ -0,0 +1,266 @@
+
+        return self._source
+
+    def calculate_size(self, progress: Optional[ProgressReporter]):
```
It looks that `progress` variable is not used here.

owtaylor replied:
Yeah, it's not used. I was basically keeping things parallel with download() in case we want add progress for this operation in the future. But I can remove it.

owtaylor commented:
Pushed a new version with the previous fixup squashed and a new fixup. (If that looks good, feel free to push a squash yourself, if that's easier.)

jkonecny12 commented:
I see that you have broken tests because you have changed API of one of the functions. This needs to get fixed.
I can try to take a look on that to help you.

jkonecny12 commented:
For RHEL-10 we need to backport also https://github.com/rhinstaller/anaconda/pull/5493 which is dependency of this change.

owtaylor commented:
> I see that you have broken tests because you have changed API of one of the functions. This needs to get fixed.
> I can try to take a look on that to help you.

If you could, that would be great!

jkonecny12 commented:
Found issues:
- I'm missing a lot of logging which makes debugging harder
- on `url` KS command use-case it seems that Flatpak size is not calculated to required space at all
- there is bug in the code for the CalculateSize call which is causing Anaconda to crash (flatpak/installation.py:70) - fixed already on line 46
- during the installation it will find firefox flatpak and start size calculation but size is 0 for installed and download and it doesn't look like the firefox flatpak is installed at all

I'll continue on this tomorrow.

Should I change your existing commits here or add a new ones to the PR for future squashing? I guess it depends on how much are you going to focus on this near term?

owtaylor commented:
> Found issues:
> 
>     * I'm missing a lot of logging which makes debugging harder

No matter how much logging I add, I usually I never have the right bit of logging for the particular bug ...

>     * on `url` KS command use-case it seems that Flatpak size is not calculated to required space at all

I'd need to know the details of the test setup. 

>     * there is bug in the code for the CalculateSize call which is causing Anaconda to crash (flatpak/installation.py:70) - fixed already on line 46

That was introduced in the last fixup when I removed the ProgressReporter argument.
 
>     * during the installation it will find firefox flatpak and start size calculation but size is 0 for installed and download and it doesn't look like the firefox flatpak is installed at all

Again, not really sure. Maybe the metadata isn't right and it doesn't think it needs to install anything.

> I'll continue on this tomorrow.
> 
> Should I change your existing commits here or add a new ones to the PR for future squashing? I guess it depends on how much are you going to focus on this near term?

You should do what you find most convenient for you - I can always interdiff as necessary.  I had no immediate plans to work on this, but if things are not working, I'm happy to collaborate on debugging them (and/or we can pull in Sebastian Wick)

jkonecny12 commented:
@owtaylor so I was finally able to get to this. The issue is that all flatpaks are skipped from size calculation (probably also installation) because expanded flatpak list is lowercase but refs used to compare are CamelCase.
```
app/org.mozilla.Firefox/x86_64/stable 
``` 
in expanded refs list is
```
app/org.mozilla.firefox/x86_64/stable
```

I'm fine to fix this but I wonder if the correct solution is just `lower()` or there is something else wrong.

jkonecny12 commented:
After putting everything to lower the flatpaks are getting downloaded during installation and size is calculated (I guess correctly) but these are not installed. Will investigate more.

jkonecny12 commented:
That explains why it is downloaded (Anaconda code) but not installed. Thanks for info.

jkonecny12 commented:
I'm adding two commits.
- first to add missing call to really calculate size of the Flatpaks
- add some debugging logs so it is reasonable to test

About the first one, the current implementation is not great. Ideally, there is a setup source with task https://github.com/rhinstaller/anaconda/blob/main/pyanaconda/modules/payloads/source/source_base.py#L104 which is called when payload is ready to setup all the sources. During this time it should download everything set mount points and basically all the non internal logic. However, looking on the code I think it could be used even when these are not directly sources, which I'm still convinced they should be. I'll look in the code and try to improve that commit without too much changes.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py:
```diff
@@ -0,0 +1,274 @@
+                          self._source_repository.url)
+                self._source = FlatpakStaticSource(self._source_repository,
+                                                   relative_path="Flatpaks")
+            else:
+                _, remote_url = conf.payload.flatpak_remote
+                log.debug("Using Flatpak registry source: %s", remote_url)
+                self._source = FlatpakRegistrySource(remote_url)
```
Also AFAIK you need to authenticate to Flatpak registry, however, I don't think this code is in Anaconda or should be in Anaconda at all. Flatpak should be able to automatically attach to RHSM subscription of the system otherwise it will be quite amount of logic here which we don't want to maintain. Also it would be harder for user to use.

jkonecny12 replied:
The CDN code is not finished it needs more polishing but that is about separate PR. I think we can ignore it right now.

jkonecny12 commented:
I think this might be the final solution to resolve issue with downloading flatpak metadata out of the task. I'll discuss this in the team to get feedback.

rvykydal commented on pyanaconda/modules/payloads/payloads.py:
```diff
@@ -238,7 +232,7 @@ def teardown_with_tasks(self):
 
-        if self._flatpak_side_payload:
-            tasks += self._flatpak_side_payload.tear_down_with_tasks()
+        if self.active_payload.side_payload:
```
Shouldn't we check for active_payload here as well (or return based on the preivous condition?, or rather add the check at the beginning as install_with_tasks() and some other methods?

rvykydal commented on pyanaconda/modules/payloads/payload/payload_base_interface.py:
```diff
@@ -50,6 +51,19 @@ def Type(self) -> Str:
+    def SidePayload(self) -> ObjPath:
+        """Side payload attached to this payload.
+
+        This payload will be started after the main payload is started.
```
Is it really "after the main payload is *started*"? Or rather handled / finished / installed ?

jkonecny12 replied:
It should be always "after" the main payload. At least in theory the point here is that side payload is somehow dependent on the main payload.

rvykydal replied:
I mean if it shouldn't rather be after the main payload finished (instead of started, ie running the main and side payload is serial, not concurrent).

jkonecny12 replied:
hmm, this could be tricky. Let's try to re-phrase.

jkonecny12 replied:
I've tried to improve the documentation strings for side payloads in general.

jkonecny12 commented:
I fixed existing tests and rebased the code to resolve test failures on the new containers.

jkonecny12 commented:
/kickstart-test --testtype smoke

jkonecny12 commented:
/kickstart-test --testtype payload

jkonecny12 commented:
Existing tests seems to be fixed. I'm running kickstart tests and web ui tests to verify we didn't break anything.
Now, I'll start working on covering new code with tests.

jkonecny12 commented:
Update:
* Fix a lot of minor issues
* Fix issue with supported sources
* Extract utilities from `source.py` to separate `utils.py`
* Add tests for flatpak module, interface, tasks and utils

Still missing `FlatpakManager` and `source.py` for test coverage. After this is resolved I'll probably do some squashing of commits.

jkonecny12 commented:
This should be the final version covered by tests. I'll move this to rhel-10 branch now to test it there.

jkonecny12 commented:
/kickstart-test --testtype smoke

jkonecny12 commented:
/kickstart-test --testtype payload

jkonecny12 commented:
/kickstart-test --testtype smoke

jkonecny12 commented:
/kickstart-test --testtype payload

jkonecny12 commented:
/packit build

jkonecny12 commented:
Add tests to cover payloads module as this code is used by all payload modules.

jkonecny12 commented:
/kickstart-test --testtype smoke

jkonecny12 commented:
/kickstart-test --testtype payload

jkonecny12 commented:
Still need more tests to cover DNF payload changes.

jkonecny12 commented:
/kickstart-test --testtype payload

jkonecny12 commented:
/kickstart-test --testtype smoke

jkonecny12 commented:
/kickstart-test --testtype payload

M4rtinK commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,435 @@
+    def _blob_url(self, digest):
+        if not digest.startswith("sha256:"):
+            raise RuntimeError("Only SHA-256 digests are supported")
+        return self._url + "/blobs/sha256/" + digest[7:]
```
This is really low level & should at least point to a specification that clearly shows what elements there are and why we are fine with the ones after this index. :P

M4rtinK commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,435 @@
+        expanded = self._expand_refs(refs)
+
+        for image in self._images:
+            if image.ref not in expanded:
```
Again, please note why we don't care about expanded refs here.

jkonecny12 replied:
Fixed.

M4rtinK commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,435 @@
+        else:
+            tag = "latest"
+
+        url_pattern = "{}/index/static?label:org.flatpak.ref:exists=1&architecture={}&tag={}"
```
This should be a constant.

jkonecny12 replied:
Fixed.

M4rtinK commented on pyanaconda/modules/payloads/payload/payload_base.py:
```diff
@@ -59,6 +63,29 @@ def type(self):
+        The side payload is a payload attached to this payload. It can be configured by this
+        payload.
+
+        This side payload calls will be automatically queued for some DBus API in Payloads module.
```
What does "some DBus API" mean in this context ?

jkonecny12 replied:
It depends on the specific API. It doesn't give sense to do this for each so it is documented for each method directly.

M4rtinK commented on pyanaconda/modules/payloads/payload/payload_base_interface.py:
```diff
@@ -50,6 +54,22 @@ def Type(self) -> Str:
+        The side payload is a payload attached to this payload. It can be configured by this
+        payload.
+
+        This side payload calls will be automatically queued for some DBus API in Payloads module.
```
Same here - "some DBus API" is not clear IMHO.

M4rtinK commented on pyanaconda/modules/payloads/payload/payload_base_interface.py:
```diff
@@ -78,6 +78,9 @@ def DefaultSourceType(self) -> Str:
 
         :return: a string representation of a source type
         """
+        if self.implementation.default_source_type is None:
```
Shouldn't we at least log a warning if we do get there & add a FIXME docstring if this really is temporary ?

M4rtinK commented on pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py:
```diff
@@ -50,6 +50,12 @@
 __all__ = ["FlatpakManager"]
 
 
+# FIXME: Workaround for https://issues.redhat.com/browse/RHEL-85624 - remove when fixed
```
Any idea if this was perhaps resolved in upstream in the meantime ?

jkonecny12 replied:
It wasn't. They know about it but it is not a priority for now as the workaround works right now and the focus is get this out.

M4rtinK replied:
OK, thanks for checking - is there perhaps an upstream bug we could link from there for future reference ?

M4rtinK commented on pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py:
```diff
@@ -233,21 +239,35 @@ def install(self, progress: ProgressReporter):
+                # If we're installing from a local source, we need to disable
+                # loading the list of current Flatpaks from the network - we
+                # want to install what was in the install image even if it is
+                # out of date.
```
Maybe we should include a justification why we want that ? Its a bit counter intuitive, as for both Fedora mirrors and Red Hat CDN or Satellite we do default to fetching the latest (and presumably more secure) RPMs from the network even if a local DVD source is available.

M4rtinK commented on pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py:
```diff
@@ -261,6 +281,29 @@ def _create_flatpak_transaction(self, installation):
 
         return transaction
 
+    # FlatpakTransaction.set_no_pull() does not leave sideload
```
This sounds like a bug - is there an upstream bugreport for this we could link from there ?

jkonecny12 replied:
It is and it is filed already. Mentioned above and adding link to the comment.

M4rtinK commented on pyanaconda/modules/payloads/payload/flatpak/flatpak.py:
```diff
@@ -19,14 +19,17 @@
 from pyanaconda.modules.payloads.base.utils import calculate_required_space
 from pyanaconda.modules.payloads.constants import PayloadType
-from pyanaconda.modules.payloads.payload.flatpak.flatpak_interface import FlatpakInterface
+from pyanaconda.modules.payloads.payload.flatpak.flatpak_interface import (
```
Can't this be squashed with some existing commit ?

M4rtinK reviewed (changes requested):
So while this seems to be going roughly in the right direction but I think there is quite a lot of improvement, mainly on the docstring side, to make the code much more maintainable & easier to debug going forward. Even just the nomenclature used by Flatpak should be very verbosely documented in the code so we can learn to work with it over time like we did with RPM, tarballs, Live installs, etc. :)

jkonecny12 commented on pyanaconda/modules/payloads/payload/dnf/dnf.py:
```diff
@@ -392,6 +392,16 @@ def calculate_required_space(self):
         return required_space.get_bytes()
 
+    def needs_flatpak_side_payload(self):
+        return True
```
Sounds like an interesting question. The main reason of this code is be able to enable Flatpak side-payload to another payloads in the future. I don't think it was designed to be able to disable this.

Using configuration file sounds interesting but if we don't have a use for that currently it would be only a dead code which I would like to avoid to have. We can implement it if there will be use-case for this.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/flatpak.py:
```diff
@@ -0,0 +1,136 @@
+# to the target system installation. Since we use the Flatpak API
+# in process, we need to do this by modifying the environment before
+# we start any threads. Setting these variables will be harmless if
+# we aren't actually using Flatpak.
```
This env variable change is for running process only. You are not able to inject variables to parent process in general.

However, extending the comment to make it obvious.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py:
```diff
@@ -0,0 +1,274 @@
+                          self._source_repository.url)
+                self._source = FlatpakStaticSource(self._source_repository,
+                                                   relative_path="Flatpaks")
+            else:
```
All our configuration fields have to have default value in anaconda.conf. Parsing validation is already handled here:
https://github.com/owtaylor/anaconda/blob/flatpak-preinstallation/pyanaconda/core/configuration/payload.py#L123

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py:
```diff
@@ -0,0 +1,274 @@
+        if not sources:
+            return
+
+        source = sources[0]
```
Adding comment to explain this. In short repository `0` is base repo not additional so Flatpak repo should be next to it.

jkonecny12 replied:
I have idea for fixing this and it would work in a way that Flatpaks would have correct own sources which will be created by the "main" payload as DNF, so this logic wouldn't be part of Flatpak itself.

That will fix other issues around this logic. But that is for future change.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,433 @@
+        :param download_location: path to location for temporary downloads
+        :param progress: used to report progress of the download
+        :returns sideload location, including the transport (e.g. oci:<path>), or None
+        """
```
Seems to be outdated.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py:
```diff
@@ -0,0 +1,274 @@
+
+        source = sources[0]
+
+        if isinstance(source, RepositorySourceMixin):
```
Fixed

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py:
```diff
@@ -0,0 +1,274 @@
+        """Sets a location that can be used for temporary download of Flatpak content.
+
+        :param path: parent directory to store downloaded Flatpak content
+           (the download should be to a subdirectory of this path)
```
This part is about finding Flatpaks for the installation but not really to download that. That happens in the method below.

jkonecny12 replied:
This seems to be outdated comment as the directory set here already have a special suffix. Removing this part.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py:
```diff
@@ -0,0 +1,274 @@
+    def calculate_size(self):
+        """Calculate the download and install size of the Flatpak content.
+
+        :param progress: used to report progress of the operation
```
Correct - removing.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py:
```diff
@@ -0,0 +1,274 @@
+
+        :param progress: used to report progress of the operation
+
+        This is only needed if Flatpak can't install the content directly.
```
It is necessary when you have a source which is not supported by Flatpak which happens for a custom repository on HTTP/FTP etc... If the source is Flatpak repository or local offline Flatpaks you don't need this.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py:
```diff
@@ -0,0 +1,274 @@
+        installation = self._create_flatpak_installation()
+        self._transaction = self._create_flatpak_transaction(installation)
+
+        if self._collection_location:
```
I've added explanation as comment on variable declaration.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py:
```diff
@@ -0,0 +1,274 @@
+        try:
+            self._progress = progress
+            self._transaction.run()
+        except GError as e:
```
Why would you want to do this? All the installation errors are fatal so I don't see much of a benefit in having it more granular.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py:
```diff
@@ -0,0 +1,274 @@
+        return transaction
+
+    def _operation_started_callback(self, transaction, operation, progress):
+        """Start of the new operation.
```
I don't think it can be explained much more than it is operation in a Transaction set but I tried :).

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/installation.py:
```diff
@@ -0,0 +1,144 @@
+
+        if os.path.exists(path):
+            log.info("Removing existing package download location: %s", path)
+            shutil.rmtree(path)
```
Not sure I follow you? This is the cleanup task which removes locally downloaded installation source which is not necessary anymore and would take space on the installed system. Am I missing something or was this review comment done on an outdated commit?

jkonecny12 replied:
I see your confusion now. This code happens before we start the download. It will make sure that there are no leftovers from before if someone would kill the Anaconda and rerun (or maybe other cases). The actual download is done after this call.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,435 @@
+
+_CONTAINER_ARCH_MAP = {
+    "x86_64": "amd64",
+    "aarch64": "arm64"
```
I think so but @owtaylor will know better than me.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,435 @@
+def _get_container_arch():
+    """Architecture name as used by docker/podman"""
+    arch = get_arch()
+    return _CONTAINER_ARCH_MAP.get(arch, arch)
```
In any case this code should work AFAIK the installation of Flatpaks is just skips if Flatpak source is missing which would be the first point where it will skip.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,435 @@
+
+
+class SourceImage(ABC):
+    """Representation of a single image of a FlatpakSource."""
```
I'm not 100% sure why it is called `Image` but as I understand it, it is basically remote Flatpak source which we need to download by FTP/HTTP manually as this is unsupported for Flatpaks.

@owtaylor could you confirm this?

jkonecny12 replied:
Add your explanation to the code. Thanks!

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,435 @@
+
+    @property
+    def ref(self) -> Optional[str]:
+        """Flatpak reference for the image, or None if not a Flatpak"""
```
Because we are taking URL from the base repository so on the place of Flatpak could be anything unsupported.

@owtaylor am I correct?

jkonecny12 replied:
Add comment to clarify.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,435 @@
+        """All images in the source, filtered for the current architecture."""
+
+    def _expand_refs(self, refs: List[str]) -> List[str]:
+        """Expand the list of refs to be in full form and include any dependencies."""
```
If you check the method you will see that the only call is in `canonicalize_flatpak_ref` where this expansion is documented already. I do not want to duplicate the documentation.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,435 @@
+                    if runtime_ref not in result:
+                        result.append(runtime_ref)
+            except (NoSectionError, KeyError):
+                log.debug("Skipping source image %s: broken metadata", image.ref)
```
Agree that these should be warnings or errors. Fixed.

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,435 @@
+
+        with self._downloader() as downloader:
+            for image in self._images:
+                if image.ref in expanded_refs:
```
Solved.

jkonecny12 commented on docs/release-notes/flatpak-inital-support.rst:
```diff
@@ -0,0 +1,22 @@
+:Description:
+    This is initial support for Flatpak preinstall feature. This feature allows marking flatpaks
+    for the installation from the target system payload. The benefit of this functionality is that
+    it is attached to the original payload (DNF packages, bootc container images…).
```
Not used anywhere right now.

jkonecny12 commented:
**Updated:**
- Rebased
- Fixed notes from @M4rtinK from above (all the fixes are in the new commits so it's easier to see the change)


Also I found out that I'm not reasonably able to test this on Fedora with the copr repository. Because on Fedora the expectation is that Anaconda will use official Fedora container repository to download Flatpaks, however, in the code this logic is started with closest mirror or CDN. Nor of these two can be specified from kickstart.
On Fedora (and maybe on RHEL too), I think we should try base URL for Flatpak source and if that doesn't work it should fallback to Repository. However, that would add another logic on top of this already quite huge PR.

So let's just merge this with unit tests support and leave the rest for RHEL / when it will be really used in Fedora.

jkonecny12 commented:
/kickstart-test --testtype smoke

owtaylor commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,435 @@
+
+
+class StaticSourceImage(SourceImage):
+    """One image of a FlatpakStaticSource."""
```
```
class SourceImage(ABC):
    """Representation of a single image of a FlatpakSource."""
class FlatpakStaticSource(FlatpakSource):
    """Flatpak images stored in a OCI image layout, either locally or remotely
class StaticSourceImage(SourceImage):
    """One image of a FlatpakStaticSource."""
```

So it's "one image stored in an OCI image layout, either locally or remotely" - if that's not clear, it's probably better to extend the doc comments for SourceImage and FlatpakStaticSource to explain what's not clear (e.g., what is an image, what is an image layout) rather than putting more docs here.

(This is private API in this file between the FlatpakSource base class and it's subclasses, which is why the class names don't start with Flatpak. I'm not sure what the general convention is for that within anaconda.)

jkonecny12 replied:
It's obvious from the directory structure that this is Flatpak so we should be fine with the name here.

jkonecny12 replied:
I think I extended the SourceImage doc string already with explanation of what is the OCI is. That should hopefully be enough for better understanding.

owtaylor commented on pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py:
```diff
@@ -0,0 +1,274 @@
+        if self._collection_location:
+            self._transaction.add_sideload_image_collection(self._collection_location, None)
+
+        self._transaction.add_sync_preinstalled()
```
Your comment looks accurate. The upstream PR got replaced:  https://github.com/flatpak/flatpak/pull/6116 

The doc comment there is:

```
/**
 * flatpak_transaction_add_sync_preinstalled:
 * @self: a #FlatpakTransaction
 * @error: return location for a #GError
 *
 * Adds preinstall operations to this transaction. This can involve both
 * installing and removing refs, based on /etc/preinstall.d contents and what
 * the system had preinstalled before.
 *
 * Returns: %TRUE on success; %FALSE with @error set on failure.
 */
```

owtaylor commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,435 @@
+    """Split off a collection ID, and add architecture if unspecified
+
+    Turn "org.fedoraproject.Stable:app/org.example.Foo//stable" into
+    ("org.fedoraproject.Stable", "app/org.example.Foo/amd64/stable")
```
Generally Flatpak allows  you to omit the architecture to mean "current architecture".

This allows:
```
Provides: flatpak-preinstall(app/org.mozilla.firefox//stable)
```

to say "this package will preinstall Firefox for the current architedcture) rather than having to substitute in the architecture in the metatdata. But there's something weird here:

 * The RHEL 10 `redhat-flatpak-preinstall-firefox-0.1-5.el10.noarch.rpm` has that with the omitted architecture
 * A Flatpak ref should have the *flatpak* arch in it: e.g., the `org.flatpak.ref` labels in https://flatpaks.redhat.io/rhel/index/static?label:org.flatpak.ref:exists=1&tag=latest use x86_64
 * But the code in `canonicalize_flatpak_ref()` is `calling get_container_arch()` not `get_arch()`, so will canonicalize that to `app/org.mozilla.firefox/amd64/stable` - which shouldn't match anything in the registry/image layout.

On other hand, this has been tested to work, both last fall, and a few months ago.... how did that work?

jkonecny12 commented on pyanaconda/modules/payloads/payload/flatpak/source.py:
```diff
@@ -0,0 +1,435 @@
+        # This is more accurate than using the org.flatpak.download-size label,
+        # because further processing of the image might have recompressed
+        # the layer using different settings.
+        return sum(int(layer["size"]) for layer in self.manifest_json["layers"])
```
I see, thanks for information. So or so it should be rare and for exception.

jkonecny12 commented:
**Updated**:
* Add better explanation for `source.py` classes
* Drop https://github.com/rhinstaller/anaconda/pull/6056/commits/3d3329fa1bf5f2114b210aba3aec2474e4d33896 as the "fix" wasn't correct
* Add explanation of architecture conversions

jkonecny12 commented:
**Update:**
* Fix test after the commit drop.

Small change visible in https://github.com/rhinstaller/anaconda/compare/cf8dea810097f766d18415aa0eb1446a255cc3e7..541e8ac16cc6e418d5f810d95cdcd1ed4fad3c4c
